include/odp/api/spec/reassembly.h

```diff
@@ -0,0 +1,120 @@
+ *
+ */
+typedef struct odp_reass_capability_t {
+	/** Reassembly offload for both IPv4 and IPv6 packets */
+	odp_bool_t ip;
+
+	/** Reassembly offload for IPv4 packets only */
+	odp_bool_t ipv4;
+
+	/** Reassembly offload for IPv6 packets only */
+	odp_bool_t ipv6;
+
+	/** Maximum time in ns that a fragment can wait in the reassembly
+	 *  offload for the arrival of further fragments.
+	 */
+	uint64_t max_wait_time;
+
+	/** Maximum number of fragments that can be reassembled */
+	uint16_t max_num_frags;
+
+} odp_reass_capability_t;
+
+/**
+ * Fragment reassembly configuration
+ *
+ * Configure inline fragment reassembly offload support. Fragment
+ * reassembly offload can be enabled in IPSEC and PKTIN operations.
+ *
+ * When the offload is enabled, fragments will be delayed for a specified time
+ * period to allow reassembly.
+ *
+ * Reassembly result will be delivered to the application through an ODP packet
+ * with reassembly metadata. odp_packet_reass_status() can be used to query if
+ * reassembly has been attempted and if reassembly was successfully comnpleted.
+ *
+ * In case of successful reassembly, the reassembled packet is delivered
+ * to the receiver as a regular ODP packet as if the reassembled packet
+ * was received as such. When reassembly is enabled in pktio, it will be
+ * be attempted before other offloads such as packet parsing, inline IPsec and
+ * classification.
+ *
+ * In case of failed reassembly, the result is delivered to the application
+ * as a special packet that does not contain valid packet data. Such a
+ * packet can be used to get information of the incomplete reassembly
+ * so that the application can try to retry or continue the reassembly.
+ * See odp_packet_reass_partial_state().
+ *
+ * Reassembly may not complete if not all fragments were received in time but
+ * also for packet parsing difficulty, fragment overlap, resource shortage or
+ * other reasons. In such cases, application may receive packets with reassembly
+ * status as either ``ODP_PACKET_REASS_NONE`` or ``ODP_PACET_REASS_INCOMPLETE``.
+ *
+ * This structure is used only for configuration, not for capability
+ * query even though it is indirectly contained in odp_pktio_capability_t.
+ * The content of odp_pktio_capability_t.config.reassembly written by
+ * odp_pktio_capability() is undefined. Reassembly capabilities of a pktio
+ * can be checked through odp_pktio_capability_t.reassembly.
+ *
+ * @see odp_packet_reass_status(), odp_packet_reass_partial_state()
+ */
+typedef struct odp_reass_config_t {
+	/** Attempt inline reassembly of IPv4 packets. Disabled by default */
+	odp_bool_t en_ipv4;
+
+	/** Attempt inline reassembly of IPv6 packets. Disabled by default */
+	odp_bool_t en_ipv6;
+
+	/** Maximum time in ns that a fragment may wait in the reassembly
+	 *  offload for the arrival of further fragments. The value may
+	 *  not exceed the max_wait_time capability.
+	 *
+	 *  Default value is 0.
+	 */
+	uint64_t max_wait_time;
```

Comment:
Since 0 is not used in practice, it could be defined as an implementation defined default value (e.g. implementation selects a value that best fits HW).

Reply:
Do you recommend documenting this or should I just drop the line which states default value as 0?

Reply:
If you do not document it but just drop the line, then zero would mean zero, not implementation defined value, and the value set by odp_ipsec_config_init() and odp_pktio_config_init() would be left undefined.

So yes, the behaviour should be documented or otherwise it does not exist.

Reply:
Agreed. So I'll add the following text to replace,

"Default value is implementation defined"

Reply:
No, that does not have the same meaning. The default value shall be defined by the the API spec to be zero. And the API spec shall define that value zero does not mean zero nanoseconds but implementation-defined wait time.

Reply:
I totally misunderstood the comment then.

I can update the text as,

"Default value is 0 and would indicate usage of implementation defined max_wait_time"

Or do you have any suggestion for the text?

Reply:
I would rather have two sentences because the two things (default value set by the init function and the meaning of zero in the config calls) are basically unrelated.

So, I would make it something like this:
```
        /** Maximum time in ns that a fragment may wait in the reassembly
         *  offload for the arrival of further fragments. The value may
         *  not exceed the max_wait_time capability. Zero value means
         *  implementation-defined maximum wait time.
         *
         *  Default value is 0.
         */
```

Reply:
Updated in v10.